### ISSUE TITLE
Numpy.einsum is leaking memory

### ISSUE THREAD

**Issue by gsimbr** [numpy/numpy#10788]
I have written a piece of code which relies on numpy.einsum quite heavily and is running over a longer period of time.

I have realized that the memory consumption for the python process is monotonically increasing although the consumption should be constant as all variables are constantly overwritten (loop over timesteps).

I have been able to reproduce the behaviour with the following minimal example:
```
  import numpy as np
  nt = int(1e9)
  q = np.random.random((3, 3, 2))
  d = np.random.random((2, 2))
  for idx in xrange(nt):
      b = np.einsum("qia, ab, qjb->ji", q, d, q)
```

After ~3 minutes the memory consumption of the respective Python process was already >200MB.

I haven't been able to track the memory leakage down with tools like meliae or pympler. 

I am using Ubuntu 16.04.4 LTS. I have been able to reproduce this with both a numpy in Python 2.7.12 (default interpreter) and Python 2.7.14 by Anaconda (numpy version was 1.14 in both cases).

Help and ideas would be much appreciated.

**Comment by seberg** [comment]
Not sure if there was something fixed, but if you are not on 1.14.2, try adding `optimize=False` to the einsum call. Or just upgrade, at least I cannot reproduce it on the same python with 1.14.2 right now.

**Comment by gsimbr** [comment]
I've updated to numpy 1.14.2, the memory consumption is at a stable level now. Thank you so much for the hint, I've invested quite some time in debugging already!

I think this issue can be closed, as numpy v1.14.2 doesn't show this behaviour. 

I don't know the policies however, whether this needs to be related to the relevant fix or issue which solved this problem.

**Comment by seberg** [comment]
Since I also could not reproduce it and the einsum stuff has been in a bit flux, I think it should be fine and I will close this. If you want to be quite sure for you, maybe try with `optimize=True` on the updated numpy. Since the default was changed back to False becuase of issues (possibly including this one).
